File: 被列覆盖的最多行数.cpp
class Solution {
public:
	int maximumRows(vector<vector<int>>& matrix, int numSelect) {
        int mask[15], suf[15];
		int m = matrix.size(), n = matrix[0].size();
        int ans = m;
        numSelect = n - numSelect;
        function<void(int, int, int)> dfs = [&](int i, int cnt, int cover) -> void {
            if (cnt == numSelect) {
                // __builtin_popcount 函数统计在二进制下 1 的个数
                ans = min(ans, __builtin_popcount(cover));
                return;
            }
            if (cnt + n - i == numSelect) {
                dfs(n, numSelect, cover | suf[i]);
            } else {
                dfs(i + 1, cnt + 1, cover | mask[i]);
                dfs(i + 1, cnt, cover);
            }
        };
        
		for (int j = 0; j < n; ++j) {
			mask[j] = 0;
			for (int i = 0; i < m; ++i) {
                mask[j] |= matrix[i][j] << i;
            }
		}
		suf[n - 1] = mask[n - 1]; 
		for (int i = n - 2; i >= 0; --i) {
            suf[i] = suf[i+1] | mask[i];
        }
		dfs(0, 0, 0);
		return m - ans;
	}
};

File: 被列覆盖的最多行数.py
class Solution:
    def maximumRows(self, matrix: List[List[int]], numSelect: int) -> int:
        mask, suf = [0 for _ in range(15)], [0 for _ in range(15)]
        m, n = len(matrix), len(matrix[0])
        ans, numSelect = m, n - numSelect

        def dfs(i, cnt, cover):
            if cnt == numSelect:
                nonlocal ans
                ans = min(ans, cover.bit_count())
                return
            if cnt + n - i == numSelect:
                dfs(n, numSelect, cover | suf[i])
            else:
                dfs(i + 1, cnt + 1, cover | mask[i])
                dfs(i + 1, cnt, cover)
        
        for j in range(n):
            mask[j] = 0
            for i in range(m):
                mask[j] |= matrix[i][j] << i

        suf[n - 1] = mask[n - 1]
        for i in range(n - 2, -1, -1):
            suf[i] = suf[i + 1] | mask[i]
        
        dfs(0, 0, 0)
        return m - ans
